runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/PuzzleSolverOverlay.java

```diff
@@ -157,79 +157,67 @@ public Dimension render(Graphics2D graphics, Point parent)
 
 				if (config.displaySolution())
 				{
-					int i = 0;
-					int lastBlankX = 0;
-					int lastBlankY = 0;
+					int markerSize = 16;
 
-					// First find the current blank tile position
-					for (int j = 0; j < itemIds.length; j++)
+					int endIndex = Math.min(solutionIndex + 5, solution.length);
+					for (int i = solutionIndex + 1; i < endIndex; ++i)
 					{
-						if (itemIds[j] == BLANK_TILE_VALUE)
-						{
-							lastBlankX = j % PuzzleSolver.DIMENSION;
-							lastBlankY = j / PuzzleSolver.DIMENSION;
-							break;
-						}
-					}
-
-					for (Integer futureMove : nextMoves)
-					{
-						int blankX = futureMove % PuzzleSolver.DIMENSION;
-						int blankY = futureMove / PuzzleSolver.DIMENSION;
-
-						int xDelta = blankX - lastBlankX;
-						int yDelta = blankY - lastBlankY;
-
-						BufferedImage arrow;
-						if (xDelta > 0)
-						{
-							arrow = getRightArrow();
-						}
-						else if (xDelta < 0)
-						{
-							arrow = getLeftArrow();
-						}
-						else if (yDelta > 0)
-						{
-							arrow = getDownArrow();
-						}
-						else
-						{
-							arrow = getUpArrow();
-						}
+						int blankX = solution[i] % PuzzleSolver.DIMENSION;
+						int blankY = solution[i] / PuzzleSolver.DIMENSION;
 
 						int x = puzzleBoxLocation.getX() + blankX * PUZZLE_TILE_SIZE
-								+ PUZZLE_TILE_SIZE / 2 - arrow.getWidth() / 2;
+								+ PUZZLE_TILE_SIZE / 2 - markerSize / 2;
 
 						int y = puzzleBoxLocation.getY() + blankY * PUZZLE_TILE_SIZE
-								+ PUZZLE_TILE_SIZE / 2 - arrow.getHeight() / 2;
-
-						OverlayUtil.renderImageLocation(graphics, new net.runelite.api.Point(x, y), arrow);
-
-						lastBlankX = blankX;
-						lastBlankY = blankY;
-
-						if (++i == 3)
-						{
-							break;
-						}
+								+ PUZZLE_TILE_SIZE / 2 - markerSize / 2;
+						graphics.setColor(Color.YELLOW);
+						graphics.fillOval(x, y, markerSize, markerSize);
+						markerSize -= 3;
 					}
 				}
 			}
 		}
+		return null;
+	}
 
-		if (solver == null || cachedItems == null || (!shouldCache && !Arrays.equals(cachedItems, itemIds)))
+	private void startSolver(int[] tiles)
+	{
+		solver = new PuzzleSolver(tiles);
+		new Thread(solver).start();
+		prevTiles = tiles;
+	}
+
+	private void updateSolutionPosition(int[] currentTiles)
```

Comment:
Please add comments/documentation on what the purpose of this method is/what it does.
It looks like you're resetting the solver after updating the solution position, so why would you need to consider the previous steps? Why not just always look at the current puzzle state and get the 5 next steps, or not reset the puzzle state?

Reply:
I made it check backwards as well because i noticed that the tiles sometimes move back again if you are laggy. This might just be graphically though. I'll look into it next time i get a puzzle.